plutus-core/plutus-core/src/PlutusCore/Constant/Typed.hs

```diff
@@ -457,39 +457,163 @@ type HasConstant term = (AsConstant term, FromConstant term)
+built-in function then is
+
+    idList :: forall a. Opaque term [a] -> Opaque term [a]
+
+Now the 'Opaque' is explicit and the 'TypeScheme' inference machinery needs to go under it in order
+to instantiate @a@. Which now does not get instantiated to an 'Opaque' as before, since we're
+already inside an 'Opaque' and can just use @Var0@ directly. So @idList@ elaborates to
+
+    idList :: Opaque term [Var0] -> Opaque term [Var0]
+
+Now let's make up some syntax for annotating contexts so that it's clear what's going on:
+
+    idList @Type |
+        :: (@Type | Opaque term (@Rep | [Var0]))
+        -> (@Type | Opaque term (@Rep | [Var0]))
+
+'@ann |' annotates everything to the right of it. The whole thing then reads as
+
+1. a builtin is always defined in the Type context
+2. @->@ preserves the Type context, i.e. it accepts it and passes it down to the domain and codomain
+3. @Opaque term@ switches the context from Type to Rep, i.e. it accepts the Type context, but
+creates the Rep context for its argument that represents a Plutus type
+
+So why the distinction?
+
+The difference between the Rep and the Type contexts that we've seen so far is that in the Rep
+context we don't need any @Opaque@, but this is a very superficial reason to keep the distinction
+between contexts, since everything that is legal in the Type context is legal in the Rep context
+as well. For example we could've elaborated @idList@ into a bit more verbose
+
+    idList :: Opaque term [Opaque term Var0] -> Opaque term [Opaque term Var0]
+
+and the world wouldn't end because of that, everything would work correctly.
+
+The opposite however is not true: certain types that are legal in the Rep context are not legal in
+the Type one and this is the reason why the distinction exists. The simplest example is
+
+    id :: Var0 -> Var0
+
+@Var0@ represents a Plutus type variable and it's a data family with no inhabitants, so it does not
+make sense to try to unlift a value of that type.
+
+Now let's say we added a @term@ argument to @Var0@ and said that when @Var0 term@ is a @GHC.Type@,
+it has a @term@ inside, just like 'Opaque'. Then we would be able to unlift it, but we also have
+things like @TyAppRep@, @TyForallRep@ and that set is open, any Plutus type can be represented
+using such combinators and we can even name particular types, e.g. we could have @PlcListRep@,
+so we'd have to special-case @GHC.Type@ for each of them and it would be a huge mess.
+
+So instead of mixing up types whose values are actually unliftable with types that are only used
+for type checking, we keep the distinction explicit.
+
+The barrier between Haskell and Plutus is the barrier between the Type and the Rep contexts and
+that barrier must always be some explicit type constructor that switches the context from Type to
+Rep. We've only considered 'Opaque' as an example of such type constructor, but we also have
```

Comment:
`Opaque :: BuiltinType -> ArbitraryType`?

Reply:
`Opaque` works in the opposite direction! It takes a representation of a Plutus type (so it may have type-level lambdas for example, although we don't provide them currently, but we provide `TyForallRep`, which is very similar) and constructs a liftable/unliftable type out of that. And in most cases `Opaque` is indexed by a type representing a Plutus type variable, so it's not about built-in types at all.